packages/odyssey-react/src/components/Radio/RadioButton.module.scss

```diff
@@ -0,0 +1,184 @@
+/*!
+ * Copyright (c) 2018-present, Okta, Inc. and/or its affiliates. All rights reserved.
+ * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
+ *
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *
+ * See the License for the specific language governing permissions and limitations under the License.
+ */
+
+/* stylelint-disable no-descending-specificity, max-nesting-depth */
+
+.radio {
+  @include is-visually-hidden;
+
+  &:checked {
+    + .label {
+      &::before {
+        border-color: $color-primary-base;
+        background-color: $color-primary-base;
+      }
+
+      &::after {
+        background: $white;
+      }
+    }
+  }
+
+  &:focus,
+  &.focus {
+    + .label {
+      &::before {
+        @include outline($focus-ring-primary, 2px);
+
+        border-color: $color-primary-base;
+      }
+    }
+  }
+
+  &:disabled {
+    + .label {
+      color: $text-body;
+      cursor: not-allowed;
+
+      &::before {
+        border-color: $border-input-disabled;
+        background-color: $bg-input-disabled;
+      }
+    }
+
+    &:checked {
+      + .label {
+        &::before {
+          border-color: $color-primary-light;
+          background-color: $color-primary-light;
+        }
+
+        &::after {
+          background-color: $white;
+        }
+      }
+    }
+  }
+
+  &:invalid,
+  &.invalid {
+    &:hover:not(:checked) + .label::before {
+      border-color: $color-danger-dark;
+    }
+
+    + .label {
+      color: $text-danger;
+
+      &::before {
+        border-color: $color-danger-base;
+      }
+    }
+
+    &:checked {
+      + .label {
+        &::before {
+          background-color: $color-danger-base;
+        }
+      }
+    }
+
+    &:focus {
+      + .label {
+        &::before {
+          @include outline($focus-ring-danger, 2px);
+
+          border-color: $color-danger-base;
+        }
+      }
+    }
+
+    &:disabled {
+      + .label {
+        color: $text-danger;
+
+        &::before {
+          border-color: $color-danger-base-disabled;
+          background: $white;
+        }
+      }
+
+      &:hover {
+        + .label {
+          &::before {
+            border-color: $color-danger-base-disabled;
+          }
+        }
+      }
+
+      &:checked {
+        + .label {
+          &::before {
+            border-color: $color-danger-base-disabled;
+            background-color: $color-danger-base-disabled;
+          }
+
+          &::after {
```

Comment:
these hyper nested selectors are kinda wild one day it might be nice to pull this apart a bit to use a flatter application of classes with less specificity

Reply:
It's a thorny problem with form elements specifically around states and valid state transitions. This particular rule assumes that the element can enter a state that is simultaneously invalid, disabled, and checked. While it may be possible to enter that state, it seem impossible to exit.

Starting with a high level discussion about what are valid state transitions could help simplify this. eg: if disabled an element can never be invalid.